keepassxc-browser/content/keepassxc-browser.js

```diff
@@ -806,35 +810,63 @@ kpxc.enablePasskeys = function() {
     passkeys.src = browser.runtime.getURL('content/passkeys.js');
     document.documentElement.appendChild(passkeys);
 
-    document.addEventListener('kpxc-passkeys-request', async (ev) => {
-        if (ev.detail.action === 'passkeys_create') {
-            const publicKey = kpxcPasskeysUtils.buildCredentialCreationOptions(ev.detail.publicKey);
-            logDebug('publicKey', publicKey);
+    let lifetimeTimer;
 
-            const ret = await sendMessage('passkeys_register', [ publicKey, window.location.origin ]);
-            if (ret) {
-                if (ret.response && ret.response.errorCode) {
-                    const errorMessage = await sendMessage('get_error_message', ret.response.errorCode);
-                    kpxcUI.createNotification('error', errorMessage);
-                }
+    const startTimer = function() {
+        lifetimeTimer = setTimeout(() => {
+            throw new DOMException('lifetimeTimer has expired', 'NotAllowedError');
+        }, PASSKEYS_LIFETIMER_TIMEOUT);
+    };
 
-                const responsePublicKey = kpxcPasskeysUtils.parsePublicKeyCredential(ret.response);
-                kpxcPasskeysUtils.sendPasskeysResponse(responsePublicKey);
-            }
-        } else if (ev.detail.action === 'passkeys_get') {
-            const publicKey = kpxcPasskeysUtils.buildCredentialRequestOptions(ev.detail.publicKey);
-            logDebug('publicKey', publicKey);
+    const stopTimer = function() {
+        if (lifetimeTimer) {
+            clearTimeout(lifetimeTimer);
+        }
+    };
 
-            const ret = await sendMessage('passkeys_get', [ publicKey, window.location.origin ]);
-            if (ret) {
-                if (ret.response && ret.response.errorCode) {
-                    const errorMessage = await sendMessage('get_error_message', ret.response.errorCode);
+    const letTimerRunOut = function(errorCode) {
+        return errorCode === PASSKEYS_WAIT_FOR_LIFETIMER || errorCode === PASSKEYS_CREDENTIAL_IS_EXCLUDED;
```

Comment:
https://www.w3.org/TR/2019/REC-webauthn-1-20190304/#sec-make-credential-privacy <- this is what I was thinking when not returning an excluded credential early. If it's possible to sniff quickly if there's a credential existing for a site, wouldn't it be better to let the extension timeout? The user is still shown a notification about an excluded credential, but the Relying Party doesn't know it.

Reply:
Yeah, I think I misunderstood when the `PASSKEYS_CREDENTIAL_IS_EXCLUDED` code was used when I wrote that comment, I thought it was used when the user had consented to reveal the credential's existence - since you're never asking for consent in the excluded case, then waiting for the timeout is the correct approach.